datumaro/components/project.py

```diff
@@ -2,486 +2,2481 @@
+            if link:
+                os.link(src, dst)
+            else:
+                shutil.copy(src, dst, follow_symlinks=True)
+
+        src = self.obj_path(obj_hash)
+        if osp.isfile(src):
+            _copy_obj(src, dst_dir, link=allow_links)
+            return
+
+        src += self.DIR_HASH_SUFFIX
+        if not osp.isfile(src):
+            raise UnknownRefError(obj_hash)
+
+        with open(src) as f:
+            src_meta = json.load(f)
+        for entry in src_meta:
+            _copy_obj(self.obj_path(entry['md5']),
+                osp.join(dst_dir, entry['relpath']), link=allow_links)
+
+    def remove_cache_obj(self, obj_hash: str):
+        src = self.obj_path(obj_hash)
+        if osp.isfile(src):
+            os.remove(src)
+            return
+
+        src += self.DIR_HASH_SUFFIX
+        if not osp.isfile(src):
+            raise UnknownRefError(obj_hash)
+
+        with open(src) as f:
+            src_meta = json.load(f)
+        for entry in src_meta:
+            entry_path = self.obj_path(entry['md5'])
+            if osp.isfile(entry_path):
+                os.remove(entry_path)
+
+        os.remove(src)
+
+class Tree:
+    # can be:
+    # - attached to the work dir
+    # - attached to a revision
+
+    def __init__(self, project: 'Project',
+            config: Optional[TreeConfig] = None,
+            rev: Optional[str] = None):
+        assert isinstance(project, Project)
+        assert not rev or project.is_ref(rev), rev
+
+        if not isinstance(config, TreeConfig):
+            config = TreeConfig(config)
+        if config.format_version != 2:
+            raise ValueError("Unexpected tree config version '%s', expected 2" %
+                config.format_version)
+        self._config = config
+
+        self._project = project
+        self._rev = rev
+
+        self._sources = ProjectSources(self)
+        self._targets = ProjectBuildTargets(self)
+
+    def save(self):
+        self.dump(self._config.config_path)
+
+    def dump(self, path):
+        os.makedirs(osp.dirname(path), exist_ok=True)
+        self._config.dump(path)
 
     @property
-    def sources(self):
+    def sources(self) -> ProjectSources:
         return self._sources
 
-    def _save_branch_project(self, extractor, save_dir=None):
-        if not isinstance(extractor, Dataset):
-            extractor = Dataset.from_extractors(
-                extractor) # apply lazy transforms to avoid repeating traversals
+    @property
+    def build_targets(self) -> ProjectBuildTargets:
+        return self._targets
+
+    @property
+    def config(self) -> Config:
+        return self._config
+
+    @property
+    def env(self) -> Environment:
+        return self._project.env
+
+    @property
+    def rev(self) -> str:
+        return self._rev
+
+    def make_dataset(self, target: Optional[str] = None) -> Dataset:
+        if not target:
+            target = 'project'
```

Comment:
Seems like you could just make `'project'` the default value of `target`.

Reply:
It can come from the CLI or from an API call as empty kwarg. I'd like to cover such uses too.